transports/noise/src/io/framed.rs

```diff
@@ -106,270 +133,70 @@ impl<T> NoiseFramed<T, snow::HandshakeState> {
         })?;
 
         let dh_remote_pubkey = PublicKey::from_slice(dh_remote_pubkey)?;
+        let codec = Codec::new(self.session.into_transport_mode()?);
 
-        let io = NoiseFramed {
-            session: self.session.into_transport_mode()?,
-            io: self.io,
-            read_state: ReadState::Ready,
-            write_state: WriteState::Ready,
-            read_buffer: self.read_buffer,
-            write_buffer: self.write_buffer,
-            decrypt_buffer: self.decrypt_buffer,
-        };
-
-        Ok((dh_remote_pubkey, Output::new(io)))
+        Ok((dh_remote_pubkey, codec))
     }
 }
 
-/// The states for reading Noise protocol frames.
-#[derive(Debug)]
-enum ReadState {
-    /// Ready to read another frame.
-    Ready,
-    /// Reading frame length.
-    ReadLen { buf: [u8; 2], off: usize },
-    /// Reading frame data.
-    ReadData { len: usize, off: usize },
-    /// EOF has been reached (terminal state).
-    ///
-    /// The associated result signals if the EOF was unexpected or not.
-    Eof(Result<(), ()>),
-    /// A decryption error occurred (terminal state).
-    DecErr,
-}
+impl asynchronous_codec::Encoder for Codec<snow::HandshakeState> {
+    type Error = io::Error;
+    type Item<'a> = &'a proto::NoiseHandshakePayload;
 
-/// The states for writing Noise protocol frames.
-#[derive(Debug)]
-enum WriteState {
-    /// Ready to write another frame.
-    Ready,
-    /// Writing the frame length.
-    WriteLen {
-        len: usize,
-        buf: [u8; 2],
-        off: usize,
-    },
-    /// Writing the frame data.
-    WriteData { len: usize, off: usize },
-    /// EOF has been reached unexpectedly (terminal state).
-    Eof,
-    /// An encryption error occurred (terminal state).
-    EncErr,
-}
+    fn encode(&mut self, item: Self::Item<'_>, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        let mut buf = Vec::with_capacity(item.get_size());
```

Comment:
Not sure if it is worth it but with this simplified design, it would be quite trivial to only allocate this `Vec` once when we instantiate `Codec` and always write to that. I think the underlying implementation should grow the buffer as needed and we could then pass on a slice of length `self.buffer[..item.get_size()]`.

Thoughts?

Reply:
Yh that sounds fine. I'm thinking of changing the currently named `write_buffer` to `encrypt_buffer` since that's what it's really used for, and then using `write_buffer` to refer to the buffer that holds the raw serialized payload in this case

Reply:
Renaming that is a good idea!